### PR TITLE
Document Collective SLIC Functions

### LINKED ISSUE #524: Clearly indicate which slic and lumberjack functions use collectives
The Slic and Lumbejack API docs should clearly indicate which slic functions/macros can use MPI collectives and which ones are always safe to call on specific ranks.

This could leak to Slic's `LogStream` derived types, such as `LumberjackStream`.

In particular, `slic::flush()` is a collective operation.

1. Make comment in the doxygen
2. Make a table in sphinx of calls which are collective, making it clear to users which ones have to called on all ranks

### PR DESCRIPTION
**Pull request by bmhan12**
This PR updates ReadTheDocs and doxygen to document which SLIC macros/functions are collective operations.
Resolves #524, relates to #26 discussion.

RTD of this branch:
https://axom.readthedocs.io/en/docs-han12-slicmpi/index.html

Table of collective SLIC macros:
https://axom.readthedocs.io/en/docs-han12-slicmpi/axom/slic/docs/sphinx/sections/appendix.html#collective-slic-macros

Slic Doxygen link:
https://axom.readthedocs.io/en/docs-han12-slicmpi/doxygen/html/namespaceaxom_1_1slic.html

Slic Macros Doxygen Link:
https://axom.readthedocs.io/en/docs-han12-slicmpi/doxygen/html/slic__macros_8hpp.html

### CODE REVIEW

File: src/axom/slic/docs/sphinx/sections/architecture.rst [hunk 1]
```diff
@@ -150,7 +155,7 @@ The list of keywords is summarized in the table below.
 | **<RANK>**          | The MPI rank that emmitted the message. Only applicable|
```
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:155]:
```suggestion
| **<RANK>**          | The MPI rank that emitted the message. Only applicable|
```
Please check and fix "emmit" elsewhere as well.

File: src/axom/slic/core/Logger.hpp [hunk 2]
```diff
@@ -173,6 +173,13 @@ class Logger
    * \param [in] filter_duplicates optional parameter that indicates whether
    * duplicate messages resulting from running in parallel will be filtered out.
    * Default is false.
+   * \note When used within an MPI distributed environment, and:
```
Comment (rhornung67) [inline src/axom/slic/core/Logger.hpp:176]: @bmhan12 since the comments you added to each of these methods is essentially the same, I think it would be more effective/impactful if the comment appeared once in a "prologue" and state that it applies to each of these methods. Specifically, use method grouping syntax in Doxygen like we do in SIdre, for example: https://github.com/LLNL/axom/blob/develop/src/axom/sidre/core/Group.hpp#L129

Also, I would move the part about the methods being collective operations and that all ranks must call them to the front before the qualifications based on message level.

File: src/axom/slic/docs/sphinx/sections/getting_started.rst [hunk 3]
```diff
@@ -55,6 +55,14 @@ an application must first specify an output destination and optionally,
 prescribe the format of the log messages. These steps are demonstrated in
 the following sections.
 
+.. warning::
+
+   If you forget to initialize Slic, Slic will call ``slic::initialize()``,
```
Comment (rhornung67) [inline src/axom/slic/docs/sphinx/sections/getting_started.rst:60]:
```suggestion
   If you do not initialize Slic, Slic will call ``slic::initialize()``,
```

File: src/axom/slic/docs/sphinx/sections/getting_started.rst [hunk 4]
```diff
@@ -112,6 +120,11 @@ This indicates that all log messages that are *debug* or higher
 are captured otherwise, the messages are ignored. Since *debug* is the lowest
 severity level, all messages will be captured in this case.
 
+.. warning::
+
+   No messages will be logged if you forget to call ``slic::setLoggingMsgLevel()``.
```
Comment (rhornung67) [inline src/axom/slic/docs/sphinx/sections/getting_started.rst:125]:
```suggestion
   No messages will be logged if you do not call ``slic::setLoggingMsgLevel()``.
```

File: src/axom/slic/interface/slic.hpp [hunk 5]
```diff
@@ -168,6 +168,13 @@ void addStreamToAllMsgLevels(LogStream* ls);
  * \param [in] filter_duplicates optional parameter that indicates whether
  * duplicate messages resulting from running in parallel will be filtered out.
  * Default is false.
+ * \note When used within an MPI distributed environment, and:
```
Comment (rhornung67) [inline src/axom/slic/interface/slic.hpp:171]: Same comments that I made earlier about grouping the methods with Doxygen syntax and preface them with a single comment about their collective nature.

File: src/axom/slic/interface/slic_macros.hpp [hunk 6]
```diff
@@ -20,8 +20,12 @@
  * \def SLIC_ERROR( msg )
  * \brief Logs an error and aborts the application.
  * \param [in] msg user-supplied message
- * \note The SLIC_ERROR macro is always active.
  * \warning This macro calls processAbort().
+ * \note The SLIC_ERROR macro is always active.
```
Comment (rhornung67) [inline src/axom/slic/interface/slic_macros.hpp:24]: Similar comment about grouping methods in Doxygen and using single comment that applies to all.

**Review by rhornung67 — CHANGES_REQUESTED** [review]
@bmhan12 I made several comments suggesting what I think is a better way to group and document the methods. Most users tend to not read the details of lengthy comments on individual methods. If the main point is stated once and is clear that it applies to all grouped methods, it is easier for users to understand I think

File: src/axom/slic/docs/sphinx/sections/architecture.rst (hunk 1 above)
Comment (rhornung67) [reply src/axom/slic/docs/sphinx/sections/architecture.rst:155]: Could be referring to Emmitt Smith? 😄

File: src/axom/slic/core/Logger.hpp [hunk 7]
```diff
@@ -166,20 +166,26 @@ class Logger
    */
   LogStream* getStream(message::Level level, int i);
 
+  ///@{
```
Comment (rhornung67) [inline src/axom/slic/core/Logger.hpp:169]: 👍

**Review by rhornung67 — APPROVED** [review]
Looks good @bmhan12 Thanks!  Others should review before merging to ensure all usage issues are addressed.

**Comment by agcapps** [comment]
I'll finish my review this morning.

File: src/axom/slic/core/Logger.hpp [hunk 8]
```diff
@@ -166,6 +166,19 @@ class Logger
...
+  //! \name Collective Methods
+  //!
+  //! \attention These methods are collective operations.
+  //! All ranks in the user-supplied communicator must call the method
+  //! when used within an MPI distributed environment.
+  //! Additionally, for logMessage:
```
Comment (agcapps) [inline src/axom/slic/core/Logger.hpp:175]: I like this grouping!  Thank you.  Rephrasing suggestion:
```suggestion
  //! The logMessage method is collective if:
```

File: src/axom/slic/docs/sphinx/sections/appendix.rst [hunk 9]
```diff
@@ -52,6 +52,66 @@ logging within an application.
   functionality in application specific macros to better suit the requirements
   of the application.
 
+.. _CollectiveSlicMacros:
+
+Collective Slic Macros
+^^^^^^^^^^^^^^^^^^^^^^^^^
+A subset of SLIC macros are collective operations when used with
+MPI-aware :ref:`logStream` instances such as :ref:`SynchronizedStream`
```
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/appendix.rst:60]:
```suggestion
MPI-aware :ref:`LogStream` instances such as :ref:`SynchronizedStream`
```

File: src/axom/slic/docs/sphinx/sections/appendix.rst [hunk 10]
```diff
@@ -52,6 +52,66 @@ logging within an application.
...
+or :ref:`LumberjackStream`.
+
+Additionally, macros such as ``SLIC_WARNING`` and ``SLIC_CHECK`` become collective
+operations when certain flags are toggled on or functions are called. Other macros
+such as ``SLIC_ERROR`` and ``SLIC_ASSERT`` can be made not collective when certain
+functions are called.
+
+The table below details which SLIC macros are collective:
+
+.. list-table:: Title
```
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/appendix.rst:70]:
```suggestion
.. list-table:: Macros, Collective and Non-Collective
```

Additionally, throughout the table I'd change the word "when" to "after" (or perhaps "if," but I like that one less), although that's more of a style change to add specificity.

File: src/axom/slic/docs/sphinx/sections/architecture.rst [hunk 11]
```diff
@@ -140,17 +145,17 @@ The list of keywords is summarized in the table below.
 +---------------------+--------------------------------------------------------+
 | **<MESSAGE>**       | The supplied message that is being logged.             |
 +---------------------+--------------------------------------------------------+
-| **<FILE>**          | The file from where the message was emmitted.          |
+| **<FILE>**          | The file from where the message was emitted.           |
```
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:148]: THANK YOU

File: src/axom/slic/docs/sphinx/sections/architecture.rst [hunk 12]
```diff
@@ -140,17 +145,17 @@ The list of keywords is summarized in the table below.
...
-| **<LINE>**          | The line location where the message was emmitted.      |
+| **<LINE>**          | The line location where the message was emitted.       |
 +---------------------+--------------------------------------------------------+
 | **<TAG>**           | A string tag associated with a given message, e.g., for|
 |                     | filtering during post-processing, etc.                 |
 +---------------------+--------------------------------------------------------+
-| **<RANK>**          | The MPI rank that emmitted the message. Only applicable|
+| **<RANK>**          | The MPI rank that emitted the message. Only applicable |
 |                     | when the `Axom Toolkit`_ is compiled with MPI enabled  |
 |                     | and with MPI-aware :ref:`logStream` instances, such as,|
```
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:157]:
```suggestion
|                     | and with MPI-aware :ref:`LogStream` instances, such as,|
```
The class is `LogStream`, with a capital L, not `logStream`.  There may be other instances of this error throughout the SLIC docs.

File: src/axom/slic/docs/sphinx/sections/architecture.rst [hunk 13]
```diff
@@ -286,7 +291,7 @@ Lumberjack Stream
 """""""""""""""""
 
 The :ref:`LumberjackStream`, is intended to be used with parallel MPI
```
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:293]:
```suggestion
The :ref:`LumberjackStream` is intended to be used with parallel MPI
```

File: src/axom/slic/docs/sphinx/sections/getting_started.rst [hunk 14]
```diff
@@ -177,7 +190,14 @@ below.
    registered with ``slic::setAbortFunction()``. See the `Slic Doxygen API Documentation`_
    for more details.
 
-Step 6: Finalize Slic
+.. note::
+
+   A subset of SLIC macros are collective operations when used with
+   MPI-aware :ref:`logStream` instances such as :ref:`SynchronizedStream`
```
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/getting_started.rst:196]:
```suggestion
   MPI-aware :ref:`LogStream` instances such as :ref:`SynchronizedStream`
```
There may be others I didn't catch.

File: src/axom/slic/interface/slic.hpp [hunk 15]
```diff
@@ -161,6 +161,19 @@ void addStreamToMsgLevel(LogStream* ls, message::Level level);
  */
 void addStreamToAllMsgLevels(LogStream* ls);
 
+///@{
+//! \name Collective Methods
+//!
+//! \attention These methods are collective operations.
+//! All ranks in the user-supplied communicator must call the method
+//! when used within an MPI distributed environment.
+//! Additionally, for logMessage:
```
Comment (agcapps) [inline src/axom/slic/interface/slic.hpp:170]: See my comment on Logger.hpp at (new) line 175.

File: src/axom/slic/interface/slic_macros.hpp [hunk 16]
```diff
@@ -62,10 +69,17 @@
...
-/// \name WARNING MACROS
-/// @{
+///@{
+//! \name WARNING MACROS
+//!
+//! \attention These warning macros are collective operations.
+//! All ranks in the user-supplied communicator must call the macro
+//! when used within an MPI distributed environment, and abort on warning
+//! messages is enabled for the current active logger (default is disabled
+//! for loggers)
```
Comment (agcapps) [inline src/axom/slic/interface/slic_macros.hpp:81]: This paragraph does need to be here: thank you for adding it.  Could you rephrase it? Currently, the paragraph starts out "These are collective macros."  Later it says "if abort on warning is enabled" and finished up with "default is disabled."

File: src/axom/slic/interface/slic_macros.hpp [hunk 17]
```diff
@@ -171,11 +192,22 @@
...
 
-  /// @}
+  ///@}
 
   //-----------------------------------------------------------------------------
-  /// \name DEBUG MACROS
   /// @{
+  //! \name CHECK MACROS
+  //!
+  //! \attention These check macros are collective operations.
```
Comment (agcapps) [inline src/axom/slic/interface/slic_macros.hpp:201]: See my comment on line 77.

**Review by agcapps — APPROVED** [review]
Thank you for this PR, @bmhan12 .  I gave a few suggestions, several of which are just style comments.

File: src/axom/slic/docs/sphinx/sections/appendix.rst [hunk 18]
```diff
@@ -52,6 +52,66 @@ logging within an application.
...
+     - No
+     - 
+   * - ``SLIC_ERROR``
+     - Yes
+     - Not collective when ``slic::disableAbortOnError()`` is called
+   * - ``SLIC_ERROR_IF``
+     - Yes
+     - Not collective when ``slic::disableAbortOnError()`` is called
+   * - ``SLIC_WARNING``
+     - Yes (see Note)
```
Comment (white238) [inline src/axom/slic/docs/sphinx/sections/appendix.rst:90]: I would remove all the `(see Note)` since there are notes for all of the yes's and its right there.  Though I could be convinced to add them to all of them but my preference is to just remove it.

File: src/axom/slic/interface/slic_macros.hpp [hunk 19]
```diff
@@ -111,22 +127,29 @@
...
+  //!
+  //! \attention These assert macros are collective operations.
+  //! All ranks in the user-supplied communicator must call the macro
+  //! when used within an MPI distributed environment, and abort on error
+  //! messages is enabled for the current active logger (default is enabled
+  //! for loggers)
+  //!
 
   /*!
  * \def SLIC_ASSERT( EXP )
```
Comment (white238) [inline src/axom/slic/interface/slic_macros.hpp:147]: This macro is not showing up properly in the doxygen:

https://axom.readthedocs.io/en/docs-han12-slicmpi/doxygen/html/slic__macros_8hpp.html#ad99a3eb01f4e3abe97a4ce625095f64e

Along with others.  It looks like the doxygen build needs `AXOM_DEBUG` enabled.

**Comment by white238** [comment]
Can you put something in the part where each macro is described that says this is collective in the appendix sphinx page:

https://axom.readthedocs.io/en/docs-han12-slicmpi/axom/slic/docs/sphinx/sections/appendix.html#slic-error

**Comment by white238** [comment]
Can you remove all the links to "Axom toolkit" and just say Axom in the warnings like this one:

https://axom.readthedocs.io/en/docs-han12-slicmpi/axom/slic/docs/sphinx/sections/appendix.html#slic-assert

File: src/axom/slic/docs/sphinx/sections/appendix.rst [hunk 20]
```diff
@@ -52,6 +52,66 @@ logging within an application.
...
+such as ``SLIC_ERROR`` and ``SLIC_ASSERT`` can be made not collective when certain
+functions are called.
+
+The table below details which SLIC macros are collective:
+
+.. list-table:: Macros, Collective and Non-Collective
+   :widths: 20 10 70
+   :header-rows: 1
+
+   * - Macro
```
Comment (kennyweiss) [inline src/axom/slic/docs/sphinx/sections/appendix.rst:74]: Nice table!

Since consecutive entries, e.g.  `SLIC_INFO` and `SLIC_INFO_IF` have the same values, would it make sense to combine these into a single row?

E.g. like this: https://stackoverflow.com/a/34222391

File: src/axom/slic/docs/sphinx/sections/architecture.rst [hunk 21]
```diff
@@ -88,32 +88,37 @@ This indicates that all messages with a level of severity of ``WARNING`` and
 higher will be captured, namely ``WARNING`` and ``ERROR`` messages. Thereby,
 enable the application to filter out messages with lower severity.
 
-.. _logStream:
+.. warning::
+
+   No messages will be logged if you forget to call ``slic::setLoggingMsgLevel()``.
+   All messages will be ignored.
```
Comment (kennyweiss) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:94]: Suggestion (minor):
```suggestion
   All messages will be ignored until the first call to ``slic::setLoggingMsgLevel()``
```

Comment (kennyweiss) [reply src/axom/slic/docs/sphinx/sections/architecture.rst:94]: Thanks for calling this out.

Is this really the design though? 
Seems `slic::initialize()` should default to some logging for its first logger, e.g. like logging warnings and errors.

Perhaps we should create an issue to discuss this?

File: src/axom/slic/docs/sphinx/sections/wrapping_slic_in_macros.rst [hunk 22]
```diff
@@ -68,6 +68,12 @@ These macros can then be used in the application code as follows:
    pass the ``__FILE__`` and ``__LINE__`` to the ``logMessage()`` function
    each time.
 
+.. warning::
+   Macros that use ``slic::logMessage()`` with a :ref:`logMessageLevel` of
+   ``WARNING`` or ``ERROR`` are collective operations when used with
+   MPI-aware :ref:`LogStream` instances. Consult :ref:`CollectiveSlicMacros`
+   for a list of collective Axom macros.
```
Comment (kennyweiss) [inline src/axom/slic/docs/sphinx/sections/wrapping_slic_in_macros.rst:75]: 👍 

Minor: Should this be a `note` rather than a `warning`?

File: src/axom/slic/interface/slic_macros.hpp (hunk 6 above)
Comment (kennyweiss) [reply src/axom/slic/interface/slic_macros.hpp:31]: 👍  for grouping.

I think it would also be helpful if the doxygen for the collective functions were tagged, e.g. with `\collective`, similar to how we use the `\accelerated` tags. This might make it easier for users to notice this when they are scrolling the header files rather than the generated doxygen.

**Review by kennyweiss — APPROVED** [review]
Thanks @bmhan12 -- this is a huge improvement!

Can you please go through the implementation of the functions to ensure that we're not missing any collective calls?
E.g. `slic::finalize()` ( and `Logger::finalize()`) is collective because it calls `flushStreams()` within it's implementation.
See: https://github.com/LLNL/axom/blob/608339cf0b9c48af3d302c4dcd8e947d475379e6/src/axom/slic/core/Logger.cpp#L333-L348

Also, I think it would be helpful if the doxygen for each function also included a `\collective` tag. Similarly to how the `\accelerated` tag was added in https://github.com/LLNL/axom/pull/620
This could be done in a follow-up PR (or not at all).

**Review by white238 — APPROVED** [review]

**Comment by white238** [comment]
Thanks @bmhan12 looks great!
